src/llama.cpp

```diff
@@ -11039,7 +11056,7 @@ struct llm_build_context {
                         ext_factor, attn_factor, beta_fast, beta_slow);
                 cb(Qcur, "Qcur", il);
 
-                Qcur = ggml_scale(ctx0, Qcur, 1.0f / sqrtf(float(n_embd_head_k)));
+                Qcur = ggml_scale(ctx0, Qcur, 1.0f / sqrtf(float(n_embd / n_head)));
```

Comment:
I am not sure that this change is correct, and it will affect other models.

Reply:
This is inside the build gemma2 graph right? So it may be incorrect but shouldn't affect other models

Reply:
Yes you are right, I though this part of the common `build_kv` function.

Reply:
I am not sure if this is correct, this change increases ppl of f16 9b-it @ 100 chunks from 11.9212 to 11.9850.

Reply:
Not sure if it's related, but original `num_key_value_heads` is set to 16, while gguf version has `gemma2.attention.head_count_kv = 8`

https://github.com/huggingface/transformers/blob/e65502951593a76844e872fee9c56b805598538a/src/transformers/models/gemma2/configuration_gemma2.py#L105

Reply:
It seems to be correct, because in the transformers implementation they do the following scaling:

https://github.com/huggingface/transformers/blob/e65502951593a76844e872fee9c56b805598538a/src/transformers/models/gemma2/modeling_gemma2.py#L256-L258

`self.scaling` is initialized on this line:

https://github.com/huggingface/transformers/blob/main/src/transformers/models/gemma2/modeling_gemma2.py#L202

So it uses the `query_pre_attn_scalar` value from the `config.json`:

https://huggingface.co/google/gemma-2-9b-it/blob/93be03fbe3787f19bf03a4b1d3d75d36cb1f6ace/config.json#L24

For the 9B model, this scalar value is `224` which is equal to `n_embd / n_head = 3584 / 16 = 224`, while `n_embd_head_k == 256`. Technically, it will be more correct to read this scaling value and use it directly, but it seems that for all Gemma 2 models so far, this factor is equal to `n_embd / n_head`. Maybe we just need to assert it during Python conversion